ccl/changefeedccl: TestChangefeedRandomExpressions failed

ccl/changefeedccl.TestChangefeedRandomExpressions [failed](https://mesolite.cluster.engflow.com/invocations/default/fbf2d9e2-59f1-490e-8705-b739ebbe8326?testReportRun=2&testReportShard=13&testReportAttempt=1#targets-Ly9wa2cvY2NsL2NoYW5nZWZlZWRjY2w6Y2hhbmdlZmVlZGNjbF90ZXN0) on release-25.4.4-rc @ [3b602e9ede0a2153e9d0ebd5823e9d1444e0d4fc](https://github.com/cockroachdb/cockroach/commits/3b602e9ede0a2153e9d0ebd5823e9d1444e0d4fc):


```
        UPDATE seed SET _enum = enum_range('hi'::greeting)[rowid%7];
    changefeed_test.go:1614: SELECT array_to_string(IFNULL(array_agg(distinct rowid),'{}'),'|') FROM seed WHERE (NOT (prettify_statement((pg_lsn('66DB4410/E71D251C':::PG_LSN::PG_LSN)::PG_LSN::PG_LSN || NULL::STRING)::STRING::STRING)::STRING ~ '=Z':::STRING))
    changefeed_test.go:1626: Skipping predicate (NOT (prettify_statement((pg_lsn('66DB4410/E71D251C':::PG_LSN::PG_LSN)::PG_LSN::PG_LSN || NULL::STRING)::STRING::STRING)::STRING ~ '=Z':::STRING)) because it returned no rows
    changefeed_test.go:1614: SELECT array_to_string(IFNULL(array_agg(distinct rowid),'{}'),'|') FROM seed WHERE seed._bool
    changefeed_test.go:1630: Expecting statement CREATE CHANGEFEED WITH schema_change_policy='stop' AS SELECT rowid FROM seed WHERE seed._bool to emit 45 events
    changefeed_test.go:1614: SELECT array_to_string(IFNULL(array_agg(distinct rowid),'{}'),'|') FROM seed WHERE '"':::STRING ~* version()::STRING
    changefeed_test.go:1626: Skipping predicate '"':::STRING ~* version()::STRING because it returned no rows
    changefeed_test.go:1614: SELECT array_to_string(IFNULL(array_agg(distinct rowid),'{}'),'|') FROM seed WHERE (e'\x7f9[\x02':::STRING::STRING % seed._string::STRING)::BOOL
    changefeed_test.go:1630: Expecting statement CREATE CHANGEFEED WITH schema_change_policy='stop' AS SELECT rowid FROM seed WHERE (e'\x7f9[\x02':::STRING::STRING % seed._string::STRING)::BOOL to emit 1 events
    changefeed_test.go:1614: SELECT array_to_string(IFNULL(array_agg(distinct rowid),'{}'),'|') FROM seed WHERE NULL
    changefeed_test.go:1626: Skipping predicate NULL because it returned no rows
    changefeed_test.go:1614: SELECT array_to_string(IFNULL(array_agg(distinct rowid),'{}'),'|') FROM seed WHERE false
    changefeed_test.go:1626: Skipping predicate false because it returned no rows
    changefeed_test.go:1614: SELECT array_to_string(IFNULL(array_agg(distinct rowid),'{}'),'|') FROM seed WHERE sha1(e'\\x65ce3f83cc3fc4a98e':::BYTES::BYTES)::STRING !~ NULL
    changefeed_test.go:1626: Skipping predicate sha1(e'\\x65ce3f83cc3fc4a98e':::BYTES::BYTES)::STRING !~ NULL because it returned no rows
    changefeed_test.go:1614: SELECT array_to_string(IFNULL(array_agg(distinct rowid),'{}'),'|') FROM seed WHERE (seed._bool OR ('':::STRING NOT LIKE seed._string))
    changefeed_test.go:1630: Expecting statement CREATE CHANGEFEED WITH schema_change_policy='stop' AS SELECT rowid FROM seed WHERE (seed._bool OR ('':::STRING NOT LIKE seed._string)) to emit 89 events
    changefeed_test.go:1614: SELECT array_to_string(IFNULL(array_agg(distinct rowid),'{}'),'|') FROM seed WHERE true
    changefeed_test.go:1630: Expecting statement CREATE CHANGEFEED WITH schema_change_policy='stop' AS SELECT rowid FROM seed WHERE true to emit 100 events
    changefeed_test.go:1614: SELECT array_to_string(IFNULL(array_agg(distinct rowid),'{}'),'|') FROM seed WHERE EXISTS (SELECT seed.tableoid AS "col""_152" FROM d.public.seed AS tab_54 WHERE seed._bool ORDER BY seed._interval DESC NULLS LAST)
    changefeed_test.go:1630: Expecting statement CREATE CHANGEFEED WITH schema_change_policy='stop' AS SELECT rowid FROM seed WHERE EXISTS (SELECT seed.tableoid AS "col""_152" FROM d.public.seed AS tab_54 WHERE seed._bool ORDER BY seed._interval DESC NULLS LAST) to emit 45 events
    changefeed_test.go:1633: Test tolerating create changefeed error: failed to start feed for job 0: pq: sub-query expressions not supported by CDC
    changefeed_test.go:1614: SELECT array_to_string(IFNULL(array_agg(distinct rowid),'{}'),'|') FROM seed WHERE ((false OR false) AND ('51kMq.xSSM4Sbx8.V.Y.koD.IUsPM_lr6':::LTREE::LTREE IN (SELECT NULL::LTREE AS col_173 FROM d.public.seed AS "ta""b_59")))
    changefeed_test.go:1626: Skipping predicate ((false OR false) AND ('51kMq.xSSM4Sbx8.V.Y.koD.IUsPM_lr6':::LTREE::LTREE IN (SELECT NULL::LTREE AS col_173 FROM d.public.seed AS "ta""b_59"))) because it returned no rows
    changefeed_test.go:1614: SELECT array_to_string(IFNULL(array_agg(distinct rowid),'{}'),'|') FROM seed WHERE (e'\x03{':::STRING::STRING % '9':::STRING::STRING)::BOOL
    changefeed_test.go:1626: Skipping predicate (e'\x03{':::STRING::STRING % '9':::STRING::STRING)::BOOL because it returned no rows
    changefeed_test.go:1614: SELECT array_to_string(IFNULL(array_agg(distinct rowid),'{}'),'|') FROM seed WHERE (NULL::INET << seed._inet::INET)::BOOL
    changefeed_test.go:1626: Skipping predicate (NULL::INET << seed._inet::INET)::BOOL because it returned no rows
    changefeed_test.go:1614: SELECT array_to_string(IFNULL(array_agg(distinct rowid),'{}'),'|') FROM seed WHERE (false AND EXISTS (SELECT 3909396145:::REGTYPE AS col_230, COALESCE(2611669263:::REGPROCEDURE, 2183407867:::REGPROCEDURE) AS "co'l_231" FROM d.public.seed AS tab_91 WHERE seed._string SIMILAR TO seed._string GROUP BY seed._date))
    changefeed_test.go:1626: Skipping predicate (false AND EXISTS (SELECT 3909396145:::REGTYPE AS col_230, COALESCE(2611669263:::REGPROCEDURE, 2183407867:::REGPROCEDURE) AS "co'l_231" FROM d.public.seed AS tab_91 WHERE seed._string SIMILAR TO seed._string GROUP BY seed._date)) because it returned no rows
    changefeed_test.go:1614: SELECT array_to_string(IFNULL(array_agg(distinct rowid),'{}'),'|') FROM seed WHERE (seed._string::STRING % e'\U00002603':::STRING::STRING)::BOOL
    changefeed_test.go:1626: Skipping predicate (seed._string::STRING % e'\U00002603':::STRING::STRING)::BOOL because it returned no rows
    changefeed_test.go:1614: SELECT array_to_string(IFNULL(array_agg(distinct rowid),'{}'),'|') FROM seed WHERE seed._string !~ to_char_with_style('1987-12-05':::DATE::DATE, seed._string::STRING)::STRING
    changefeed_test.go:1619: Skipping query SELECT array_to_string(IFNULL(array_agg(distinct rowid),'{}'),'|') FROM seed WHERE seed._string !~ to_char_with_style('1987-12-05':::DATE::DATE, seed._string::STRING)::STRING because error pq: unknown DateStyle parameter: 1
    changefeed_test.go:1614: SELECT array_to_string(IFNULL(array_agg(distinct rowid),'{}'),'|') FROM seed WHERE ('53.251.156.61':::INET::INET >> '8100:25f7:751e:8ee9:5861:528c:77c7:d11d/27':::INET::INET)::BOOL
    changefeed_test.go:1626: Skipping predicate ('53.251.156.61':::INET::INET >> '8100:25f7:751e:8ee9:5861:528c:77c7:d11d/27':::INET::INET)::BOOL because it returned no rows
    changefeed_test.go:1614: SELECT array_to_string(IFNULL(array_agg(distinct rowid),'{}'),'|') FROM seed WHERE ((seed._bool::BOOL NOT IN (SELECT false::BOOL AS col_306 FROM d.public.seed AS "tab""_115" ORDER BY seed._bool NULLS FIRST, seed._decimal ASC NULLS FIRST, seed._timestamp ASC NULLS FIRST, seed._bool NULLS FIRST, seed._interval DESC NULLS LAST, seed._interval NULLS FIRST, seed._date NULLS FIRST, seed._decimal NULLS LAST)) OR st_covers(seed._string::STRING, '':::STRING::STRING)::BOOL)
    changefeed_test.go:1619: Skipping query SELECT array_to_string(IFNULL(array_agg(distinct rowid),'{}'),'|') FROM seed WHERE ((seed._bool::BOOL NOT IN (SELECT false::BOOL AS col_306 FROM d.public.seed AS "tab""_115" ORDER BY seed._bool NULLS FIRST, seed._decimal ASC NULLS FIRST, seed._timestamp ASC NULLS FIRST, seed._bool NULLS FIRST, seed._interval DESC NULLS LAST, seed._interval NULLS FIRST, seed._date NULLS FIRST, seed._decimal NULLS LAST)) OR st_covers(seed._string::STRING, '':::STRING::STRING)::BOOL) because error pq: error parsing EWKT: syntax error: unexpected NUM at line 1, pos 0
        LINE 1: 2
                ^
    changefeed_test.go:1614: SELECT array_to_string(IFNULL(array_agg(distinct rowid),'{}'),'|') FROM seed WHERE (NOT EXISTS (SELECT seed.tableoid AS "co-l_309", '':::VOID AS "col\v_310"))
    changefeed_test.go:1626: Skipping predicate (NOT EXISTS (SELECT seed.tableoid AS "co-l_309", '':::VOID AS "col\v_310")) because it returned no rows
    changefeed_test.go:1614: SELECT array_to_string(IFNULL(array_agg(distinct rowid),'{}'),'|') FROM seed WHERE st_issimple(st_mpolyfromtext(seed._string::STRING, seed._int4::INT8)::GEOMETRY::GEOMETRY)::BOOL
    changefeed_test.go:1619: Skipping query SELECT array_to_string(IFNULL(array_agg(distinct rowid),'{}'),'|') FROM seed WHERE st_issimple(st_mpolyfromtext(seed._string::STRING, seed._int4::INT8)::GEOMETRY::GEOMETRY)::BOOL because error pq: error parsing EWKT: syntax error: unexpected NUM at line 1, pos 0
        LINE 1: 1
                ^
    changefeed_test.go:1614: SELECT array_to_string(IFNULL(array_agg(distinct rowid),'{}'),'|') FROM seed WHERE (true AND (((seed._inet::INET << seed._inet::INET)::BOOL AND (seed._string::STRING % to_char(transaction_timestamp():::TIMESTAMP::TIMESTAMP::TIMESTAMP, seed._string::STRING)::STRING::STRING)::BOOL) = (seed._inet::INET << '4140:bec:23bf:7011:9e19:1ca:a0ff:1c3b/86':::INET::INET)::BOOL))
    changefeed_test.go:1630: Expecting statement CREATE CHANGEFEED WITH schema_change_policy='stop' AS SELECT rowid FROM seed WHERE (true AND (((seed._inet::INET << seed._inet::INET)::BOOL AND (seed._string::STRING % to_char(transaction_timestamp():::TIMESTAMP::TIMESTAMP::TIMESTAMP, seed._string::STRING)::STRING::STRING)::BOOL) = (seed._inet::INET << '4140:bec:23bf:7011:9e19:1ca:a0ff:1c3b/86':::INET::INET)::BOOL)) to emit 93 events
    changefeed_test.go:1686: got 0 of 93 expected changefeed messages: operation "kafka.Next-1146725217067892737" timed out after 30s (given timeout 30s): context deadline exceeded
--- FAIL: TestChangefeedRandomExpressions/kafka (47.28s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>deadlock=true</code>
 - <code>run=2</code>
 - <code>shard=13</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #162124 ccl/changefeedccl: TestChangefeedRandomExpressions failed [A-cdc C-test-failure O-robot T-cdc branch-release-25.4 release-blocker]
- #155779 ccl/changefeedccl: TestChangefeedRandomExpressions failed [A-cdc C-test-failure O-robot T-cdc branch-release-25.2]
- #155746 ccl/changefeedccl: TestChangefeedRandomExpressions failed [A-cdc C-test-failure O-robot P-3 T-cdc branch-release-25.4.0-rc]
- #154500 ccl/changefeedccl: TestChangefeedRandomExpressions failed [A-cdc C-test-failure O-robot P-3 T-cdc branch-master]
- #147573 ccl/changefeedccl: TestChangefeedRandomExpressions failed [A-cdc C-test-failure O-robot P-3 T-cdc branch-release-25.1.7-rc]
</p>
</details>
/cc @cockroachdb/cdc
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestChangefeedRandomExpressions.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-59313

--- a/pkg/ccl/changefeedccl/changefeed_test.go
+++ b/pkg/ccl/changefeedccl/changefeed_test.go
@@ -1787,7 +1787,21 @@ func TestChangefeedRandomExpressions(t *testing.T) {
 				continue
 			}
 			whereClausesChecked[where] = struct{}{}
-			query = "SELECT array_to_string(IFNULL(array_agg(distinct rowid),'{}'),'|') FROM seed WHERE " + where
+			// Disable the optimizer and use the row query execution engine. This
+			// helps ensure we avoid cases where the oracle query has different
+			// behavior than the internal query. For example, the vectorized engine
+			// and optimizer may prevent a `false and crash()` expression from
+			// returning an error even though it will fail when the cdc query
+			// attempts to execute it.
+			//
+			// NOTE: DB is a connection pool. We set the options in the statement to
+			// ensure they are set on the connection that is actually executing the
+			// query. We can't easily use a single connection because when queries
+			// timeout, the context cancellation closes the connection.
+			query = `
+				SET testing_optimizer_disable_rule_probability = 1.0;
+				SET vectorize = off;
+				SELECT array_to_string(IFNULL(array_agg(distinct rowid),'{}'),'|') FROM seed WHERE ` + where
 			t.Log(query)
 			timeoutCtx, cancel := context.WithTimeout(context.Background(), time.Second*2)
 			rows := s.DB.QueryRowContext(timeoutCtx, query)
@@ -1819,54 +1833,11 @@ func TestChangefeedRandomExpressions(t *testing.T) {
 			}
 			err = assertPayloadsBaseErr(context.Background(), seedFeed, assertedPayloads, false, false, nil, changefeedbase.OptEnvelopeWrapped)
 			closeFeedIgnoreError(t, seedFeed)
-			if err != nil {
-				// Skip errors that may come up during SQL execution. If the SQL query
-				// didn't fail with these errors, it's likely because the query was built in
-				// a way that did not have to execute on the row that caused the error, but
-				// the CDC query did.
-				// Since we get the error that caused the changefeed job to
-				// fail from scraping the job status and creating a new
-				// error, we unfortunately don't have the pgcode and have to
-				// rely on known strings.
-				validPgErrs := []string{
-					"argument is not an object",
-					"cannot subtract infinite dates",
-					"dwithin distance cannot be less than zero",
-					"error parsing EWKB",
-					"error parsing EWKT",
-					"error parsing GeoJSON",
-					"expected LineString",
-					"geometry type is unsupported",
-					"invalid escape string",
-					"invalid input syntax for type",
-					"invalid regular expression",
-					"no locations to init GEOS",
-					"parameter has to be of type Point",
-					"regexp compilation failed",
-					"result out of range",
-					"should be of length",
-					"unknown DateStyle parameter",
-				}
-				containsKnownPgErr := func(e error) (interface{}, bool) {
-					for _, v := range validPgErrs {
-						if strings.Contains(e.Error(), v) {
-							return nil, true
-						}
-					}
-					return nil, false
-				}
-				if _, contains := errors.If(err, containsKnownPgErr); contains {
-					t.Logf("Skipping statement %s because it encountered pgerror %s", createStmt, err)
-					continue
-				}
-
-				t.Fatal(err)
-			}
+			require.NoError(t, err)
 			numNonTrivialTestRuns++
 		}
 		require.Greater(t, numNonTrivialTestRuns, 0, "Expected >0 predicates to be nontrivial out of %d attempts", n)
 		t.Logf("%d predicates checked: all had the same result in SELECT and CHANGEFEED", numNonTrivialTestRuns)
-
 	}
 
 	cdcTest(t, testFn, feedTestForceSink(`kafka`), withAllowChangefeedErr("changefeed may have parsing failure on some queries"))

